### DISCUSSION TITLE
Make Specify a Progressive Web App

### DISCUSSION THREAD

**Discussion by grantfitzsimmons** [specify/specify7#5709] [Feature Requests]
We need to finish the process of making Specify a PWA. The initial work was started with https://github.com/specify/specify7/pull/2752 with the color scheme and other info laid out in the `manifest.json` file in the `static` directory.

We need to make it install-able and do some testing on this process. Need to add the service worker and test this. On my to-do list but time is hard to come by.

> @grantfitzsimmons:
> 
> https://web.dev/codelab-make-installable/
> 
> There is some more work to be done to make this installable from what I understand. 
> 
> We need to add a service worker to this branch to enable installing the app.
> 
> https://glitch.com/edit/#!/make-it-installable?path=service-worker.js%3A1%3A0
> 
> See this example `service-worker.js`:
> 
> ```js
> const CACHE_NAME = 'offline';
> const OFFLINE_URL = 'offline.html';
> 
> self.addEventListener('install', function(event) {
>   console.log('[ServiceWorker] Install');
>   
>   event.waitUntil((async () => {
>     const cache = await caches.open(CACHE_NAME);
>     // Setting {cache: 'reload'} in the new request will ensure that the response
>     // isn't fulfilled from the HTTP cache; i.e., it will be from the network.
>     await cache.add(new Request(OFFLINE_URL, {cache: 'reload'}));
>   })());
>   
>   self.skipWaiting();
> });
> 
> self.addEventListener('activate', (event) => {
>   console.log('[ServiceWorker] Activate');
>   event.waitUntil((async () => {
>     // Enable navigation preload if it's supported.
>     // See https://developers.google.com/web/updates/2017/02/navigation-preload
>     if ('navigationPreload' in self.registration) {
>       await self.registration.navigationPreload.enable();
>     }
>   })());
> 
>   // Tell the active service worker to take control of the page immediately.
>   self.clients.claim();
> });
> 
> self.addEventListener('fetch', function(event) {
>   // console.log('[Service Worker] Fetch', event.request.url);
>   if (event.request.mode === 'navigate') {
>     event.respondWith((async () => {
>       try {
>         const preloadResponse = await event.preloadResponse;
>         if (preloadResponse) {
>           return preloadResponse;
>         }
> 
>         const networkResponse = await fetch(event.request);
>         return networkResponse;
>       } catch (error) {
>         console.log('[Service Worker] Fetch failed; returning offline page instead.', error);
> 
>         const cache = await caches.open(CACHE_NAME);
>         const cachedResponse = await cache.match(OFFLINE_URL);
>         return cachedResponse;
>       }
>     })());
>   }
> });
> ```

_Potential refactoring to consider down the road:_

> https://github.com/specify/specify7/blob/ccc6c8705a589c24f287dfb151dd9dbb72e81570/specifyweb/settings/__init__.py#L144-L147
> 
> @maxpatiiuk:
> Is this required? You might be able to achieve the same by defining a view that would serve the manifest.json page:
> 
> View:
> 
> https://github.com/specify/specify7/blob/3dced0964721218de64659e8371b1c218c6d3fef/specifyweb/urls.py#L28
> 
> it's imported here:
> 
> https://github.com/specify/specify7/blob/3dced0964721218de64659e8371b1c218c6d3fef/specifyweb/urls.py#L11
> 
> the view itself is defined like this: (https://specifydev.slack.com/archives/D02LXH9PEUU/p1672952503649319)
> 
> https://github.com/specify/specify7/blob/3dced0964721218de64659e8371b1c218c6d3fef/specifyweb/frontend/views.py#L20-L29

**Comment by maxpatiiuk** [comment]
Duplicate of #2626, but I will merge that one with this, as this issue has more details

**Comment by maxpatiiuk** [comment]
- [x] create a manifest.json file (including the ability to run specify in borderless mode and brand colors)
- [ ] configure web notifications ([Make Notifications fetching and stats reporting more efficient #2606](https://github.com/specify/specify7/issues/2606))
- [ ] service worker (for partial offline mode and cache control)
